pkg/receiver/gitproviderreceiver/internal/scraper/gitlabscraper/gitlab_scraper.go

```diff
@@ -70,56 +106,73 @@ func (gls *gitlabScraper) scrape(ctx context.Context) (pmetric.Metrics, error) {
+	for hasNextPage := true; hasNextPage; {
+		// TODO: since we pass in a context already, do we need to create a new background context?
+		// Get the next page of data
+		projects, err := getAllGroupProjects(context.Background(), graphClient, gls.cfg.GitLabOrg, projectsCursor)
 		if err != nil {
-			gls.logger.Sugar().Errorf("error getting descendant groups: %v", err)
+			gls.logger.Sugar().Errorf("error: %v", err)
 		}
-		for _, project := range projects.Group.Projects.Nodes {
-			projectList = append(projectList, project.Name)
+
+		if len(projects.Group.Projects.Nodes) == 0 {
+			err = errors.New("no GitLab projects found for the given group/org")
+			gls.logger.Sugar().Error(err)
+
+			return gls.mb.Emit(), err
 		}
-	}
 
-	// TODO: Must paginate and update query to get all branches of projects in the list
-	//for _, project := range projectList {
+		// Check if there is a next page
+		hasNextPage = projects.Group.Projects.PageInfo.HasNextPage
 
-	//	gls.logger.Sugar().Debugf("project: %v", project)
+		// Set the cursor to the end cursor of the current page
+		projectsCursor = &projects.Group.Projects.PageInfo.EndCursor
 
-	//	branches, err := getBranchNames(context.Background(), graphClient, project)
+		for _, p := range projects.Group.Projects.Nodes {
+			projectList = append(projectList, gitlabProject{
+				Name:           p.Name,
+				Path:           p.FullPath,
+				CreatedAt:      p.CreatedAt,
+				LastActivityAt: p.LastActivityAt,
+			})
+		}
+	}
 
-	//	if err != nil {
-	//		gls.logger.Sugar().Errorf("error getting branches: %v", err)
-	//	}
+	ch := make(chan branchData)
 
-	//	var branchList []string
-	//	branchList = append(branchList, branches.Project.Repository.BranchNames...)
+	var wg sync.WaitGroup
 
-	//	gls.logger.Sugar().Debugf("branch: %v", branches.Project.Repository.BranchNames)
+	// TODO: Must account for when there are more than 100,000 branch names in a project.
+	for _, project := range projectList {
+		wg.Add(1)
 
-	//	gls.mb.RecordGitRepositoryBranchCountDataPoint(now, int64(len(branchList)), project)
+		// created shadowed declaration due to loop variable because Loop variables
+		// captured by 'func' literals in 'go' statements might have unexpected values
+		project := project
+		go func() {
+			gls.getBranches(ctx, graphClient, project.Path, ch, &wg)
+		}()
+	}
 
-	//}
+	go func() {
+		wg.Wait()
+		close(ch)
+	}()
 
-	if err != nil {
-		gls.logger.Sugar().Errorf("error: %v", err)
+	for branch := range ch {
+		branchCount := int64(len(branch.BranchNames))
+
+		gls.mb.RecordGitRepositoryBranchCountDataPoint(now, branchCount, branch.ProjectPath)
+		gls.logger.Sugar().Debugf("%s branch count: %v", branch.ProjectPath, branchCount)
 	}
 
+	// record repository count metric
 	gls.mb.RecordGitRepositoryCountDataPoint(now, int64(len(projectList)))
-
 	gls.logger.Sugar().Debugf("metrics: %v", gls.cfg.Metrics.GitRepositoryCount)
 
+	gls.logger.Sugar().Infof("took %v\n", time.Since(now.AsTime()))
```

Comment:
```suggestion
```

Reply:
Ope, thanks for this! This was a holdover from me testing to see how much time go routines would save us.